common/core/web/input-processor/src/text/inputProcessor.ts

```diff
@@ -128,19 +129,25 @@ namespace com.keyman.text {
 
               let altEvent = altKey.constructKeyEvent(this.keyboardProcessor, mock, keyEvent.device);
               let alternateBehavior = this.keyboardProcessor.processKeystroke(altEvent, mock);
-              if(alternateBehavior) {
-                // TODO: if alternateBehavior.beep == true, set 'p' to 0.  It's a disallowed key sequence,
-                //       so a user should never have intended to type it.  Should probably renormalize 
-                //       the distribution afterward, though...
-                
+              
+              // If alternateBehavior.beep == true, ignore it.  It's a disallowed key sequence,
+              // so we expect users to never intend their use.
+              if(alternateBehavior && !alternateBehavior.beep) {
                 let transform: Transform = alternateBehavior.transcription.transform;
                 
                 // Ensure that the alternate's token id matches that of the current keystroke, as we only
                 // record the matched rule's context (since they match)
                 transform.id = ruleBehavior.transcription.token;
                 alternates.push({sample: transform, 'p': pair.p});
+                totalMass += pair.p;
               }
             }
+
+            // Renormalizes the distribution, as any error (beep) results
+            // will result in a distribution that doesn't sum to 1 otherwise.
+            alternates.forEach(function(alt) {
+              alt.p /= totalMass;
```

Comment:
Is there ever a possibility of totalMass == 0?

Reply:
It's the same probability of a keyboard being designed that could conceivably output a BEEP for literally every output keystroke on one layer for the same context.

So... I'm going with "practically, no."  It should take very, very poor design for that to happen.

Reply:
> It should take very, very poor design for that to happen.

So, practically guaranteed to happen at some point then? 🤣